### PR TITLE
[MXNET-982] Provide example to illustrate usage of CSVIter in C++ API

### PR DESCRIPTION
**Pull request by leleamol**
## Description ##
This PR includes an example that uses C++ API and demonstrates the usage of CSVIter.
It also includes unit test for this example and python script that converts the MNIST data to CSV format.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [y] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [y] Changes are complete (i.e. I finished coding on this PR)
- [y] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [y] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [y] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by stu1130** [comment]
Thanks for your contribution  @leleamol 
@mxnet-label-bot[pr-work-in-progress]

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    .SetParam("data_csv", "./data/mnist_test.csv")
+    .SetParam("data_shape", Shape(num_mnist_features + 1, 1))
+    .SetParam("batch_size", batch_size)
+    .SetParam("flat", 1)
+    .SetParam("shuffle",0)
+    .CreateDataIter();
+
+    auto net = mlp(layers);
+    
+    Context ctx = Context::cpu();  // Use CPU for training
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: can you make this configurable?

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    return SoftmaxOutput(outputs.back(), label);
+}
+
+
+int main(int argc, char** argv)
+{
+    const int image_size = 28;
+    const int num_mnist_features = image_size * image_size;
+    const std::vector<int> layers{128, 64, 10};
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: layers->hidden_layer_size? layers generally is referred to convolution, pooling, dense, etc.,

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            // Extract the feature data by slicing the columns 1 to 785 of the data and copy it to "X" arg.
+            reshapedData.Slice(1,(num_mnist_features + 1)).Reshape(Shape(batch_size,num_mnist_features)).CopyTo(&args["X"]);
+            
+            // Forward pass is enough as no gradient is needed when evaluating
+            exec->Forward(false);
+            acc.Update(labelData, exec->outputs[0]);
+        }
+        float duration = std::chrono::duration_cast<std::chrono::milliseconds>
+        (toc - tic).count() / 1000.0;
+        LG << "Epoch: " << iter << " " << samples/duration << " samples/sec Accuracy: " << acc.Get();
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: Can you follow the same format as the Python Examples. 
see this https://github.com/apache/incubator-mxnet/pull/12182#issuecomment-416634452

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+             */
+            NDArray reshapedData = data_batch.data.Reshape(Shape((num_mnist_features + 1),batch_size));
+            
+            // Extract the label data by slicing the first column of the data and copy it to "label" arg.
+            reshapedData.Slice(0,1).Reshape(Shape(batch_size)).CopyTo(&args["label"]);
+            
+            // Extract the feature data by slicing the columns 1 to 785 of the data and copy it to "X" arg.
+            reshapedData.Slice(1,(num_mnist_features + 1)).Reshape(Shape(batch_size,num_mnist_features)).CopyTo(&args["X"]);
+            
+            // Compute gradients
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: move this comment to the next line(forward does not compute gradients)

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    .SetParam("flat", 1)
+    .SetParam("shuffle",0)
+    .CreateDataIter();
+
+    auto net = mlp(layers);
+    
+    Context ctx = Context::cpu();  // Use CPU for training
+    
+    std::map<std::string, NDArray> args;
+    args["X"] = NDArray(Shape(batch_size, num_mnist_features), ctx);
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: why "X"? we use "data" for the input layer

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    return SoftmaxOutput(outputs.back(), label);
+}
+
+
+int main(int argc, char** argv)
+{
+    const int image_size = 28;
+    const int num_mnist_features = image_size * image_size;
+    const std::vector<int> layers{128, 64, 10};
+    const int batch_size = 100;
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: can you make all of these configurable, lets make sure we follow the same as Python Examples.

File: cpp-package/example/mlp_csv_cpu.cpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    const int max_epoch = 10;
+    const float learning_rate = 0.1;
+    const float weight_decay = 1e-2;
+
+    /*
```
Comment (nswamy) [inline cpp-package/example/mlp_csv_cpu.cpp]: nicely commented, this is very useful for examples 👍

**Comment by leleamol** [comment]
@nswamy Thanks for the review comments! I will address them and post an update.
Regarding the README, after this example, I am planning to write a README for all the examples. That file would explain what's the purpose of each example and how to run it.

Conversion from MNIST to CSV ~30 seconds . It's less than the time required to download these files.

File: cpp-package/example/mlp_csv.cpp [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
```
Comment (nswamy) [inline cpp-package/example/mlp_csv.cpp]: we [don't add Author tags](https://www.theinquirer.net/inquirer/news/1037207/apache-enforces-the-removal-of-author-tags), please remove

File: cpp-package/example/mlp_csv.cpp [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    std::string token;
+    while ((pos = hidden_units_string.find(delimiter)) != std::string::npos) {
+        token = hidden_units_string.substr(0, pos);
+        hidden_units.push_back(atoi(token.c_str()));
+        hidden_units_string.erase(0, pos + delimiter.length());
+    }
+    hidden_units.push_back(atoi(hidden_units_string.c_str()));
+}
+
+void printUsage()
```
Comment (nswamy) [inline cpp-package/example/mlp_csv.cpp]: do you want to pull out the arg parsing to a different file to keep this clean? may be you can generalize arg parsing later as you add new examples to C++

File: cpp-package/example/unittests/unit_test_mlp_csv.sh [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+# This file is a unit test for mlp_csv.cpp example in 'example' directory.
+# The file
+#    1. Downloads the MNIST data,
+#    2. Converts it into CSV format.
+#    3. Runs the mlp_csv example and ensures that the accuracy is more than expected.
+#
+
+#!/bin/bash
+
+set -e # exit on the first error
```
Comment (nswamy) [inline cpp-package/example/unittests/unit_test_mlp_csv.sh]: I am wondering why chose shell script for this? this example as Cpp, Py and sh..may be just use python here, WDYT?

File: cpp-package/example/mlp_csv.cpp (hunk 8 above)
Comment (leleamol) [reply cpp-package/example/mlp_csv.cpp]: I will remove it. I had seen it in the other mlp examples hence added it.

File: cpp-package/example/mlp_csv.cpp (hunk 9 above)
Comment (leleamol) [reply cpp-package/example/mlp_csv.cpp]: I would defer pulling out or refactoring it.  As we make more changes to these examples, we will get better idea and we can come up with more generic command line parser.

File: cpp-package/example/unittests/unit_test_mlp_csv.sh (hunk 10 above)
Comment (leleamol) [reply cpp-package/example/unittests/unit_test_mlp_csv.sh]: I chose shell script because I found it easier to invoke python and cpp programs from shell and process the output using grep.

There was no other particular reason.  We can revisit later and may be tie unit tests for all examples in single framework later.

File: cpp-package/README.md [hunk 11]
```diff
@@ -1,42 +1,21 @@
...
 
-A basic tutorial can be found at <https://mxnet.incubator.apache.org/tutorials/c++/basics.html>.
+From cpp-package/examples directory
+-  Build all examples in release mode: **make all**
+-  Build all examples in debug mode : **make debug**
 
-## Examples
+By default, the examples are build to be run on GPU.
+To build examples to run on CPU:
+- Release: **make all MXNET_USE_CPU=1**
```
Comment (nswamy) [inline cpp-package/README.md]: not all laptops/computers have GPU, can you make CPU the default.

**Comment by leleamol** [comment]
I had mistakenly added README files that were supposed to go with the new PR.
I have reverted those changes.  I noticed the travis-ci build is failing but it doesn't seem to be related to this PR.

Can this PR be merged?

File: cpp-package/example/mlp_csv.cpp [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    std::string delimiter = ",";
+    size_t pos = 0;
+    std::string token;
+    std::vector<int> hidden_units;
+    while ((pos = hidden_units_string.find(delimiter)) != std::string::npos) {
+        token = hidden_units_string.substr(0, pos);
+        hidden_units.push_back(atoi(token.c_str()));
+        hidden_units_string.erase(0, pos + delimiter.length());
+    }
+    hidden_units.push_back(atoi(hidden_units_string.c_str()));
```
Comment (larroy) [inline cpp-package/example/mlp_csv.cpp]: This is potentially undefined behaviour if hidden_units_string is not a valid int. Correctness is not clear from the immediate context.

File: cpp-package/example/mlp_csv.cpp [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+            return 0;
+        }
+        index++;
+    }
+
+    if (training_set.empty() || test_set.empty() || hidden_units_string.empty()) {
+        std::cout << "ERROR: The mandatory arguments such as path to training and test data or "
+        << "number of hidden units for mlp are not specified." << std::endl << std::endl;
+        printUsage();
+        return 0;
```
Comment (larroy) [inline cpp-package/example/mlp_csv.cpp]: we should return 1 on errors.

File: cpp-package/example/mlp_csv.cpp [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    std::vector<int> hidden_units = getLayers(hidden_units_string);
+
+    if (hidden_units.empty()) {
+        std::cout << "ERROR: Number of hidden units are not provided in correct format."
+        << "The numbers need to be separated by ','." << std::endl << std::endl;
+        printUsage();
+        return 0;
```
Comment (larroy) [inline cpp-package/example/mlp_csv.cpp]: see above.

File: cpp-package/example/mlp_csv.cpp [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    while (index < argc) {
+        if (strcmp("--train", argv[index]) == 0) {
+            index++;
+            training_set = argv[index];
+        } else if (strcmp("--test", argv[index]) == 0) {
+            index++;
+            test_set = argv[index];
+        } else if (strcmp("--epochs", argv[index]) == 0) {
+            index++;
+            max_epoch = atoi(argv[index]);
```
Comment (larroy) [inline cpp-package/example/mlp_csv.cpp]: unvalidated atoi is dangerous, strtol might be better.

File: cpp-package/example/unittests/unit_test_mlp_csv.sh (hunk 10 above)
Comment (larroy) [reply cpp-package/example/unittests/unit_test_mlp_csv.sh]: Python and check_call is more maintainable.

Comment (leleamol) [reply cpp-package/example/unittests/unit_test_mlp_csv.sh]: Sure @larroy . I will consider these suggestions when I will start implementing the test framework.

**Comment by leleamol** [comment]
The ci-build with travis is failing because it is exceeding the time allocated for the job. It doesn't seem to be related to the changes in PR.
@mxnet-jenkins @marcoabreu

**Comment by nswamy** [comment]
Please add OSX tests to travis in your next iteration.
